### ISSUE TITLE
Feature Request: Option for exit insert mode when switching to other "buffer"/pane.

### ISSUE THREAD

**Issue by jordwalke** [t9md/atom-vim-mode-plus#535]
There's an option to go back into normal mode when the editor blurs. One problem is that this fires when switching apps. This is a problem when doing something like typing in the search editor `/` and cmd+tabbing to another app to recall the word you're trying to search for, or also when quickly command+tabbing to another program while typing in insert mode to recall the word you want to type. As a general rule, I feel the best user experience is that the editor returns to normal mode when jumping to any other pane inside of the editor (including file nav or project search, or another tab/pane), but not when command+tabbing to another app. What I am suggesting is the same behavior as MacVim.

**Comment by t9md** [comment]
Could you summarize in short your expectation?
- return to normal-mode when changing active pane within atom editor
- do nothing when active app changed

Correct? Is there similar option in MacVim? Want to know where and the option name.

And you can do this for blur event

https://github.com/t9md/atom-vim-mode-plus/wiki/TIPS#escape-insert-mode-when-focus-lost

**Comment by t9md** [comment]
#337

**Comment by jordwalke** [comment]
> return to normal-mode when changing active pane within atom editor
> do nothing when active app changed

Your description is correct, but I was mistaken that MacVim has a setting for this.

But for vmp, I think it is pretty intuitive that changing focus to any new pane, should make the *previous* pane enter normal mode. I have been trying that [wiki suggestion](https://github.com/t9md/atom-vim-mode-plus/wiki/TIPS#escape-insert-mode-when-focus-lost) and it has the unintuitive behavior that I pointed out, which is why I filed this issue. I do think that if we can get the experience to be as you described (return all existing cursors in all panes to normal mode when switching active pane within atom, but not when app changed) it would be a great config option (and possibly even on by default, but not sure).

(Also, I don't know if modals like `cmd+p` should cause existing editors to return to normal mode - modals are very *temporary*, and might be canceled quickly).

**Comment by jordwalke** [comment]
I just realized that *part* of the behavior I was seeing (undesirable) (when searching `/` `cmd+tab` clears out search) was not due to that `init` snippet that I had from the wiki - it seems to be the default behavior.

**Comment by t9md** [comment]
So is this agreement?
I need spec before start implementation.

- Return to normal-mode when changing active pane within Atom editor
- Do nothing on app-switch(`cmd-tab`)
- Don't clear search-mini-editor(`/`, `?`) on app-switch

**Comment by jordwalke** [comment]
That is correct. The following might *also* be good things to consider:

- Make a configuration option to turn this behavior on/off.
- I'm not sure what should happen when `cmd+p` is pressed. 

Thanks for listening.

**Comment by t9md** [comment]
OK, will evaluate soon.

**Comment by t9md** [comment]
> Don't clear search-mini-editor(/, ?) on app-switch

For this behavior, I think the above behavior should be default. So current behavior is bad.
No need to be user-controllable, so split to other issue.

**Comment by jordwalke** [comment]
awesome.
